torch/csrc/jit/interpreter.cpp

```diff
@@ -204,24 +422,35 @@ struct InterpreterStateImpl {
   void runOneStage(
     const std::vector<at::Tensor> & inputs,
     std::vector<at::Tensor> & outputs) {
-      //std::cout << "running stage: " << current_stage << " of " << function->stages.size() << "\n";
+      // std::cout << "running stage: " << current_stage << " of " << function->stages.size() << "\n";
       JIT_ASSERT(current_stage < function->stages.size());
       auto & stage = function->stages[current_stage++];
       JIT_ASSERT((int)inputs.size() == stage.inputs.size);
       for(int i = 0; i < stage.inputs.size; i++) {
         int reg = get(stage.inputs,i);
         if(reg >= 0) { // otherwise this input is dead, and we do not store it to avoid holding the reference
-          registers[reg] = inputs[i];
+          registers.save(reg, inputs[i]);
         }
-        //std::cout << "registers[" << reg << "] = inputs[" << i << "](" << inputs[i].defined() << ")\n";
+        //std::cout << "registers[" << reg << "] = inputs[" << i << "](" << registers[reg] << ")\n";
       }
       for(auto & inst : stage.instructions) {
-        loadTensorsFromRegisters(inst.inputs, input_buffer);
+        auto & inputs = inst.inputs.values;
+        for(int i = 0; i < inputs.size; i++) {
+          int reg = get(inputs,i);
+          input_buffer.push_back(registers[reg]);
+          //std::cout << "inputs[" << i << "] = registers[" << reg << "](" << registers[reg] << ")\n";
+        }
         inst.callback(input_buffer, output_buffer);
         for(int i = 0; i < inst.outputs.size; i++) {
           int reg = get(inst.outputs,i);
-          registers[reg] = std::move(output_buffer[i]);
-          //std::cout << "registers[" << reg << "] = outputs[" << i << "](" << output_buffer[i].defined() << ")\n";
+          registers[reg] = output_buffer[i];
```

Comment:
Can't you use `.save` here to steal the references?

---

Now it seems to me that this line is leaking (since `refcounted_list` doesn't do refcounting, and you don't steal the references, `output_buffer` won't `decref` the things it holds).

Reply:
This is fine - ownership is being transferred from `output_buffer` to `register`. output_buffer won't release, but registers also wont retain. It will be clearer with better types, which I will add.